File: core/src/org/tetawex/ecf/util/FileChooser.java
package org.tetawex.ecf.util;

import java.util.Arrays;
import java.util.Comparator;

import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import org.tetawex.ecf.core.ECFGame;
import org.tetawex.ecf.screen.StyleFactory;

public class FileChooser extends Dialog {
    private FileHandle directory;
    private FileHandle file;

    private ECFGame game;

    public FileHandle getFile() {
        return file;
    }

    @Override
    public Dialog show(Stage stage) {
        return super.show(stage);
    }

    public void setDirectory(FileHandle directory) {
        if (this.directory != directory) {
            this.directory = directory;
            this.file = null;
            buildList();
        }
    }

    public void setFile(FileHandle file) {
        if (this.file != file) {
            if (this.file != null) {
                Label label = (Label) this.findActor(this.file.name());
                label.setColor(new Color(1,1,1,0.87f));
            }
            Label label = (Label) this.findActor(file.name());
            label.setColor(new Color(0.3f, 0.11f, 0.18f, 0.87f));
            this.file = file;
        }
    }

    public FileChooser(String title, WindowStyle style, ECFGame game) {
        super(title, style);
        this.game=game;
        this.getCell(getButtonTable()).expandX().fill();
        this.getButtonTable().defaults().expandX().pad(16).fill();

        this.button("Load", new FileHandle(""), StyleFactory.generateStandardMenuButtonStyle(game));
        this.button("Cancel",null, StyleFactory.generateStandardMenuButtonStyle(game));

        this.setModal(true);
    }

    public FileChooser(String title, Skin skin, String windowStyleName) {
        super(title, skin, windowStyleName);
        this.setModal(true);
    }

    private void buildList() {
        FileHandle[] files = directory.list();
        Arrays.sort(files, new Comparator<FileHandle>() {
            @Override
            public int compare(FileHandle o1, FileHandle o2) {
                if (o1.isDirectory() && !o2.isDirectory()) {
                    return -1;
                }
                if (o2.isDirectory() && !o1.isDirectory()) {
                    return +1;
                }
                return o1.name().compareToIgnoreCase(o2.name());
            }
        });
        ScrollPane pane = new ScrollPane(null, StyleFactory.generateStandardScrollPaneStyle(game));
        Table table = new Table().top().left();
        table.defaults().left();
        ClickListener fileClickListener = new ClickListener(){
            @Override
            public void clicked(InputEvent event, float x, float y) {
                Label target = (Label) event.getTarget();
                if (target.getName().equals("...")) {
                    setDirectory(directory.parent());
                } else {
                    FileHandle handle = directory.child(target.getName());
                    if (handle.isDirectory()) {
                        setDirectory(handle);
                    } else {
                        setFile(handle);
                    }
                }
            }
        };
        table.row();
        Label label = new Label("...", StyleFactory.generateStandardLabelStyle(game));
        label.setName("...");
        label.addListener(fileClickListener);
        table.add(label).expandX().pad(16).fillX();
        for (FileHandle file : files) {
            table.row();
            label = new Label(file.name(), StyleFactory.generateStandardLabelStyle(game));
            label.setName(file.name());
            label.addListener(fileClickListener);
            table.add(label).expandX().pad(16).fillX();
        }
        pane.setWidget(table);
        this.getContentTable().reset();
        this.getContentTable().add(pane).size(1200,1200);
    }

}

File: core/src/org/tetawex/ecf/util/FileChooser.kt
package org.tetawex.ecf.util

import java.util.Arrays
import java.util.Comparator

import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import org.tetawex.ecf.core.ECFGame
import org.tetawex.ecf.presentation.StyleFactory

open class FileChooser(title: String, style: Window.WindowStyle, private val game: ECFGame) : Dialog(title, style) {
    private var directory: FileHandle? = null
    private var file: FileHandle? = null

    fun getFile(): FileHandle? {
        return file
    }

    override fun show(stage: Stage): Dialog {
        return super.show(stage)
    }

    fun setDirectory(directory: FileHandle) {
        if (this.directory !== directory) {
            this.directory = directory
            this.file = null
            buildList()
        }
    }

    fun setFile(file: FileHandle) {
        if (this.file !== file) {
            if (this.file != null) {
                val label = this.findActor<Actor>(this.file!!.name()) as Label
                label.color = Color(1f, 1f, 1f, 0.87f)
            }
            val label = this.findActor<Actor>(file.name()) as Label
            label.color = Color(0.3f, 0.11f, 0.18f, 0.87f)
            this.file = file
        }
    }

    init {
        this.getCell(buttonTable).expandX().fill()
        this.buttonTable.defaults().expandX().pad(16f).fill()
        this.button("Load", FileHandle(""), StyleFactory.generateStandardMenuButtonStyle(game))
        this.button("Cancel", null, StyleFactory.generateStandardMenuButtonStyle(game))
        this.isModal = true
    }

    private fun buildList() {
        val files = directory!!.list()
        Arrays.sort(files, Comparator { o1, o2 ->
            if (o1.isDirectory && !o2.isDirectory) {
                return@Comparator -1
            }
            if (o2.isDirectory && !o1.isDirectory) {
                +1
            } else o1.name().compareTo(o2.name(), ignoreCase = true)
        })
        val pane = ScrollPane(null, StyleFactory.generateStandardScrollPaneStyle(game))
        val table = Table().top().left()
        table.defaults().left()
        val fileClickListener = object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                val target = event.target as Label
                if (target.name == "...") {
                    setDirectory(directory!!.parent())
                } else {
                    val handle = directory!!.child(target.name)
                    if (handle.isDirectory) {
                        setDirectory(handle)
                    } else {
                        setFile(handle)
                    }
                }
            }
        }
        table.row()
        var label = Label("...", StyleFactory.generateStandardLabelStyle(game))
        label.name = "..."
        label.addListener(fileClickListener)
        table.add(label).expandX().pad(16f).fillX()
        for (file in files) {
            table.row()
            label = Label(file.name(), StyleFactory.generateStandardLabelStyle(game))
            label.name = file.name()
            label.addListener(fileClickListener)
            table.add(label).expandX().pad(16f).fillX()
        }
        pane.widget = table
        this.contentTable.reset()
        this.contentTable.add(pane).size(1200f, 1200f)
    }

}
